qemu/target/i386/excp_helper.c

```diff
@@ -635,19 +635,19 @@ static int handle_mmu_fault(CPUState *cs, vaddr addr, int size,
     /* align to page_size */
     pte &= PG_ADDRESS_MASK & ~(page_size - 1);
     page_offset = addr & (page_size - 1);
-    paddr = get_hphys(cs, pte + page_offset, is_write1, &prot);
+    /* HACK allow full 64 bit mapping in u64 without paging */
+    if (env->cr[0] & CR0_PG_MASK) {
+        paddr = get_hphys(cs, pte + page_offset, is_write1, &prot);
+    } else {
+        paddr = addr;
```

Comment:
It looks like not congruent with the previous logic? At least we should have `paddr = vaddr = addr & TARGET_PAGE_MASK`.

Reply:
This is already done direct before the tlb_set_page_with_attrs() call (line 648). But if you like it better I can add this.